### ISSUE TITLE
Bimsie1AuthInterface:Login always yields TRUE

### ISSUE THREAD

**Issue by Lawri-van-Buel** [opensourceBIM/BIMserver#275]
Version: Bimserver 1.4 ([Release:1446595200000 /1.4.0-FINAL-2015-11-04](https://github.com/opensourceBIM/BIMserver/releases/tag/1.4.0-FINAL-2015-11-04))  
 
Test used: [BIMSie API Interface](https://buildingsmart.github.io/BIMSie/) -->  Bimsie1AuthInterface --> isLoggedIn  
Type: JSON
Request: 
```
{
  "request": {
    "interface": "Bimsie1AuthInterface", 
    "method": "login", 
    "parameters": {
      "username": "test",
      "password": "test"
    }
  }
}
```
Response:
```
{
  "response": {
    "result": true
  }
}
```
This is always True, regardless of the presence of a token, or a session.
I Also noticed all HTTP-Status Codes are 200, even when having an error, or when you get a not permitted error. This is not conform the [RFC-7231](https://tools.ietf.org/html/rfc7231).

**Comment by rubendel** [comment]
Will have a look at this, probably will replace this method with a call like 'isAnActualUser'.

It is conform HTTP because the actual HTTP request was successfully handled. I don't see why you would want to do some of the feedback to an API call via the HTTP status codes, and other bits via the returned JSON. The enduser is going to have to look in the returned JSON anyways. But this is a very general discussion happening everywhere on the internet, so lets not do it here.

**Comment by Lawri-van-Buel** [comment]
You are talking about an API and a REST interface. 
and when you use them there are a few standardized documentation to follow.

Status codes  are an important indicator for how to handle a request, especially to identify all cases were something went wrong. 
When an application does not communicate with status codes and automation / automated-logging is used, __ALL DATA__ must be parsed. This is not only expensive in means of computing time and memory, but also means multiple validation for the same data.
this than results in (for example) that a downloaded IFC file is read multiple times before it ever reaches the end target . many of which could be avoided.
  
As Bimserver is not a Webservice with a dedicated frontend (where using status codes can be pointless) but a generic interface. it should return meaningful return code.
 This enables machines and humans to know what is going on.

There are 3 main states to return a specific status code for:
 
 - All is ok, we use the 2xx range for this. like the [200] OK.
 - You need more rights, this is for when you need to login or need other permissions to access the resource. we use either the 401 or 403 return codes for thes.
 - An error has occurred (this is different than an authentication requirement) we use the 5xx range to indicate errors.

[RFC-7235](https://tools.ietf.org/html/rfc7235#section-3.1) 
expands RFC-7230 and RFC-73231 with authentication mechanisms and corresponding status codes.

when you return a 200  (as specified in [w3-rfc2616](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html) ) you are saying for a GET verb:  ``An entity corresponding to the requested resource is sent in the response.`` this is an invalid response when there is an error (the returned item does not contain what was requested, but an error report) , so a different status code should be used. 

the 401 and 403 return codes specificity tell the applications that the authentication is at fault here with subtle differences, ``401 Unauthorized`` is most functionally correct for the stated purpose of missing credentials. 

in short. your application does not conform to the standards. and your using assumptions you can not make for an generic interface. (e.a. there is a Human enduser).

**Comment by rubendel** [comment]
You are talking about resources and REST, but the BIMserver API is not a REST interface (and if I have a say in it, there never will be a REST interface). There are quite a few indicators for this not being REST:
- Just one URI (/json)
- Only POST requests
- It will always respond with JSON, no content-type is requested by the client

If browsers were able to talk anything other that HTTP (plain sockets for example), we would use those methods, but HTTP is the only option (aside from websockets), we just use it to tunnel the RPC calls.

I actually agree that there should be response codes (in the JSON), but that's just too much work for now to implement.

The 200, 401 and 403 HTTP status codes are pretty clear as to when to send them, but there are quite a few more, which - to be a full implementation of HTTP - should all be used. A lot of them are pretty ambigous in the context of an RPC interface.

My opinion still is that the HTTP status codes are on another level. For example, when you are  proxying requests, The 401 Unauthorized status code could indicate that a certain client IP address is not allowed to connect. How is a client going to determine whether this is a BIMserver authentication error or something else?

**Comment by coryasilva** [comment]
Technically, 401 is to say that the client is `Unauthorized`, meaning he will need to acquire a token before resubmitting the same request.  403 is `Forbidden`, meaning you do not have permission (key word) to access this resource.  So 403 should be used for the client IP address restriction.

I like adhering to standards as it helps other developers quickly understand exposed services.  Though, I do admit that it is not really that difficult to understand a non-REST HTTP service.  For example look at [Salesforce's REST api](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/intro_rest_resources.htm). 

If you had time/money/resources would you want the HTTP API to be RESTful?

**Comment by Lawri-van-Buel** [comment]
Ruben, The list you give why its not a RESTful interface is besides the point. A [RESTful](https://en.wikipedia.org/wiki/Representational_state_transfer) interface is an Architectural style. nothing more. Almost all HTTP based protocols are RESTful due to HTTP being designed (in retrospect) RESTful.

Further more the RFC I refer to are about the HTTP Protocol, not RESTful specific. ergo your assumption that you adhere to the HTTP specification is incorrect. (the fact there called HTTP states codes and not REST statues codes could have been a hint) 

A proxy should emit the 403 and 401 when the proxy is configured to do so or when the bimserver emits it.  The client is not supposed to have to care about who gives the response, a proxy should be 'transparent' enough so that the authentication mechanism works regardless of who emitted the Unauthorized / Forbidden

**Comment by rubendel** [comment]
@coryasilva Ah right, so that was a bad example. Nope, I personally don't like REST and RESTfull (there used to be a distinction between the two) for most situations. Implementing it would mean changing everything in BIMserver from RPC style to resource-style. For example, the SOAP and Protocol Buffers implementations would have no place in it. Just one simple issue I have with REST is that the resource identifiers are basically 1 dimensional (just a string of text) and even limited to about 2000 characters.

@Lawri-van-Buel So yeah we are talking about 2 different things (in a topic that has nothing to do with any of them).

For the HTTP status codes. Please tell me the response codes for:
- A programming error which results in something undefined (I'd say 500 here)
- A user error, for example getting information about a project that does not exist, in this case, the request does not come back with project information, "the returned item does not contain what was requested, but an error report". 500? 404?
- A JSON syntax error in the request

And how is the client/logging/analytics platform going to do something useful with the responses based solely on the status codes?

**Comment by Lawri-van-Buel** [comment]
Well, Personally I use the following status codes:

 - 200: Action is accepted and completed.
 - 201: New entity Created.
 - 202: request Accepted, pending action and or result.
 - 301/302/303 Redirection in various forms. (I use these only on the Web server themselves, not in Web apps. 
 - 400: Malformed or otherwise Bad request.
 - 401: When Authentication is needed.
 - 403: When more Rights / permissions are required (like a normal user tries to access an admin page)
 - 418: When I want to test  an server / service. and for ping like functionality. (of course with the RFC 2324 status message)
 - 500: Programming mistake, e.a. a mistake that should never happen. and  all other non specific errors.
 - 503: Configuration Mistakes and throttling. I use for configuration mistakes because they should be quickly resolved and discovered during the smoke test.

Status codes are great for funneling different type of errors into different actions / logs. 
Automation can even trigger automated alerts, informing operations of errors without the need for a 24/7 monitoring engineer.

You are correct in that the codes themselves do not convey enough information for automated resolving of errors. but does contain enough for the funneling / differentiation. and are great classes for statistical analysis of operation of a web service.

**Comment by berlotti** [comment]
This is a very general discussion happening everywhere on the internet, so lets not do it here.
Will have a look at this. Thank you very much for reporting it.
